### PR TITLE
Comprehensive Model, Inference, and Training Updates: Lazy Priors, Broken Powerlaw+TwoPeak, Multi-Spin Extensions, and Stable Regressor Improvements

### LINKED ISSUE #626: running parallel NumPyro chains more than number of devices on multi-gpu platforms
### Check for existing issues

- [x] Completed

### Description

See [discussion](https://forum.pyro.ai/t/how-to-run-more-chains-than-number-of-devices-in-parallel/8958) for details and answer.

### PR DESCRIPTION
**Pull request by Qazalbash**
## Related To

- #135
- #646
- closes #626

## #643

Implementation of broken powerlaw and 2 peaks, from [GWTC-4.0: Population Properties of Merging Compact Binaries](https://arxiv.org/abs/2508.18083).

## #645

What if priors depend on some runtime value of another prior? This PR introduces a feature that allows the use of priors dependent on other priors, which we refer to as Lazy Priors because they are constructed at runtime. To facilitate the functionality of Lazy Priors, we have provided the `LazyJointDistribution`.

As a simple case could be,

$$
(x_1,x_2)\sim\mathrm{Unif}(0,1)\cdot\mathrm{Unif}(0, x_1)
$$

where, $x_2$ is samples from $\mathrm{Unif}(0, x_1)$, whose upperbound is equal to $x_1$, which is being samples from a $\mathrm{Unif}(0,1)$. We can represent these interdependencies as follows in our priors,

```json
{
    "x1": {
        "dist": "Uniform",
        "high": 1.0,
        "low": 0.0
    },
    "x2": {
        "dist": "Uniform",
        "high": "x1",
        "low": 0.0
    }
}
```

> [!NOTE]
> 
> GWKokab team likes to thank Syed Mujtaba Hasan Naqvi ([@nitrodragonoid](https://github.com/nitrodragonoid)) for enhancing the dependency graph traversal (topological sort).

## #647

Injection parser for O1+O2+O3+O4a cumulative sensitivities.

## #649

This PR introduces significant improvements to the neural regressor training routine for pdet prediction.
The goal is to make training smoother, more stable, and numerically consistent while providing interpretable metrics (such as RMSE) for better model evaluation.

**Key Changes:**

	•	Optimizer: Switched from plain Adam to AdamW for improved generalization and weight regularization.
	•	Learning Rate Scheduler: Added cosine decay with warmup to gradually ramp up learning and prevent sharp loss oscillations.
	•	Gradient Clipping: Applied clip_by_global_norm to prevent gradient explosions and reduce loss spikes.
	•	Loss Function Options:
	•	Added support for Binary Cross-Entropy with logits (bce_logits) — suitable for training on [0,1] probability targets such as pdet.
	•	Retained existing MSE loss for regression-style experiments.
	•	Validation Split: Now uses a fixed random seed to ensure reproducible train/validation splits.
	•	RMSE Metric: Added Root Mean Square Error logging on the probability scale (after sigmoid), giving interpretable feedback even when BCE loss remains high.
	•	Logging: Expanded progress logging to include key training hyperparameters and evaluation results.
	•	Plots: Retained the familiar _loss.png and _total_loss.png diagnostics for consistency.
	
**Motivation:**

The previous implementation used plain Adam with a constant learning rate and unbounded gradients, which often resulted in spiky validation losses and unstable convergence.
This new setup yields:
	•	smoother learning curves,
	•	lower validation variance,
	•	improved interpretability of the final model’s performance on pdet predictions.

## #650

Refactoring the naming and flags for spin magnitudes, and implementing truncated normal distributions for aligned spins.

## #651

This PR provides a feature to pass custom estimators for poisson mean.

## #652

Continue the work of https://github.com/gwkokab/gwkokab/pull/645 for the numpyro sampler.

## #653

Apparently equations for Broken Powerlaw + Two Peak are wrong in [GWTC-4.0: Population Properties of Merging Compact Binaries](https://arxiv.org/abs/2508.18083). Paper describes the use of left truncated normal distribution as two of the three components of mixture distribution describing the primary mass, while the correct version uses truncation on both side in normal distribution.

## #655

We have replaced the iterative approach of varitaional inference with simpler expressions, that are similar to Laplace's approximations.

## #658

Implementation of solution proposed in [discussion](https://forum.pyro.ai/t/how-to-run-more-chains-than-number-of-devices-in-parallel/8958), with relevant refactoring.

## #660

Broken Powerlaw + Two Peak has been extended by providing each part its independent spin and tilt distribution. New model is called `BrokenPowerlawTwoPeakMultiSpinMultiTiltFull`.

## #663

Refactoring `RQSpline_MALA_Bundle` to optionally select Hamiltonian Monte Carlo.

`RQSpline_MALA_Bundle` and `mala_step_size` in flowMC configuration has been renamed to `bundle_config` and `step_size` respectively.

To choose between MALA and HMC, provide `local_sampler_name` in `bundle_config`; valid `local_sampler_name` are `mala` (default) and `hmc`.

`step_size` is shared between MALA and HMC and there are two more paramters `n_leapfrog`; number of steps for leapforg integrator, and `condition_matrix`; condition matrix which can either be positive scalar, vector of size equal to number of features with positive entries, or square matrix of size number of features which should be positive definite (all eigne values should be positive).

## Summary by CodeRabbit

* **New Features**
  * New two-peak broken power-law models with optional spin, tilt, eccentricity, and redshift plus CLI entry points.
  * Custom Poisson-mean estimator support and support for dependent (lazy) priors.
  * Extended injection data loader supporting O1/O2/O3/O4a formats and new HMC local sampler option.

* **Improvements**
  * Analytical likelihood now uses deterministic precision/covariance refinements with PSD safeguards.
  * Expanded, more granular spin parameterizations and consistent parameter naming.

* **Chores**
  * API surface and export cleanups, new tests, and training utilities updates.

### CODE REVIEW

**Comment by Qazalbash** [comment]
@coderabbitai full review

**Review by zeeshan5885 — APPROVED** [review]
good Job Meesum, things seems working fine. Minor changes in future improvements, but overall code is reaching to the stable position.
